### PR TITLE
feat: Turn All `Call::new_in` into `Call::new` (i.e. Stop Supporting Reentrancy)

### LINKED ISSUE #434: [Feature]: Turn All `Call::new_in` into `Call::new` (i.e. Stop Supporting Reentrancy)
### What is the feature you would like to see?

[Slack discussion](https://openzeppelin.slack.com/archives/C06R5AGFE9G/p1732632475656059).
In short, the team agreed we would not be supporting reentrant patterns.

Doing this will enable us to remove `TopLevelStorage` impls from most, if not all contracts.

TODO:

- [ ] Turn all `Call::new_in` into `Call::new` calls (wherever applicable).
- [ ] Update all functions that were `&mut self` only because of using `Call::new_in` to now be just `&self`.
- [ ] Remove `TopLevelStorage` impls wherever possible.

### Contribution Guidelines

- [X] I agree to follow this project's Contribution Guidelines

### PR DESCRIPTION
**Pull request by SarveshLimaye**
Resolves #434 

Replaced all `Call::new_in` into  `Call::new`

#### PR Checklist

- [x] Tests
- [x] Documentation

### CODE REVIEW

File: contracts/src/utils/cryptography/ecdsa.rs [hunk 1]
```diff
@@ -8,7 +8,6 @@ use alloy_primitives::{address, uint, Address, B256, U256};
 use alloy_sol_types::{sol, SolType};
 use stylus_sdk::{
     call::{self, Call, MethodError},
-    storage::TopLevelStorage,
```
Comment (0xNeshi) [inline contracts/src/utils/cryptography/ecdsa.rs:11]: Let's try removing all occurrences of `TopLevelStorage` from the whole project, theoretically we shouldn't need it now

File: contracts/src/utils/cryptography/ecdsa.rs [hunk 2]
```diff
@@ -8,7 +8,6 @@ use alloy_primitives::{address, uint, Address, B256, U256};
...
     stylus_proc::SolidityError,
 };
```
Comment (0xNeshi) [inline contracts/src/utils/cryptography/ecdsa.rs:13]: Update _CHANGELOG.md_ with a short description of your changes in this PR.

**Review by 0xNeshi — COMMENTED** [review]
Nice and clean! :rocket:
A couple more things and we're good to go

File: contracts/src/utils/cryptography/ecdsa.rs (hunk 1 above)
Comment (qalisander) [reply contracts/src/utils/cryptography/ecdsa.rs:11]: Yeah. Correct. We need implement them with these changes

Comment (SarveshLimaye) [reply contracts/src/utils/cryptography/ecdsa.rs:11]: Thanks for the review @0xNeshi @qalisander . Removed all occurrences of `TopLevelStorage`

File: contracts/src/utils/cryptography/ecdsa.rs (hunk 2 above)
Comment (SarveshLimaye) [reply contracts/src/utils/cryptography/ecdsa.rs:13]: Updated `CHANGELOG.md`

File: CHANGELOG.md [hunk 3]
```diff
@@ -18,7 +18,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Implement `MethodError` for `safe_erc20::Error`. #402
 - Use `function_selector!` to calculate transfer type selector in `Erc1155`. #417
-- Update internal functions of `Erc721` and `Erc721Consecutive` accept reference to `Bytes`. #437 
+- Update internal functions of `Erc721` and `Erc721Consecutive` accept reference to `Bytes`. #437
```
Comment (0xNeshi) [inline CHANGELOG.md:21]:
```suggestion
- Update internal functions of `Erc721` and `Erc721Consecutive` to accept a reference to `Bytes`. #437
```

File: contracts/src/token/erc1155/mod.rs [hunk 4]
```diff
@@ -786,7 +780,7 @@ impl Erc1155 {
     /// interface id or returned with error, then the error
     /// [`Error::InvalidReceiver`] is returned.
     fn _check_on_erc1155_received(
-        &mut self,
+        &self,
```
Comment (0xNeshi) [inline contracts/src/token/erc1155/mod.rs:783]: Don't forget to update the documentation above each function for which the signature has changed

**Review by 0xNeshi — COMMENTED** [review]
Much better, almost there!

File: CHANGELOG.md [hunk 5]
```diff
@@ -18,7 +18,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Implement `MethodError` for `safe_erc20::Error`. #402
 - Use `function_selector!` to calculate transfer type selector in `Erc1155`. #417
-- Update internal functions of `Erc721` and `Erc721Consecutive` accept reference to `Bytes`. #437 
+- Update internal functions of `Erc721` and `Erc721Consecutive` to accept a reference to `Bytes`. #437
+- Replace all `Call:new_in` to `Call:new` to stop supporting reentrancy. #440
```
Comment (bidzyyys) [inline CHANGELOG.md:22]: I would say it should be marked as `### Changed (Breaking)`

**Review by bidzyyys — COMMENTED** [review]
Great job @SarveshLimaye!

File: contracts/src/token/erc1155/mod.rs (hunk 4 above)
Comment (SarveshLimaye) [reply contracts/src/token/erc1155/mod.rs:783]: Thanks for the reminder, @0xNeshi. I've updated the docs for each function where the signature has changed!

File: CHANGELOG.md (hunk 5 above)
Comment (SarveshLimaye) [reply CHANGELOG.md:29]: Thanks for the review @0xNeshi @bidzyyys ! Added `### Changed (Breaking)` section

**Review by bidzyyys — COMMENTED** [review]
Looks good to me!

File: CHANGELOG.md [hunk 6]
```diff
@@ -23,7 +23,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Implement `AddAssignUnchecked` and `SubAssignUnchecked` for `StorageUint`. #418
 - Implement `MethodError` for `safe_erc20::Error`. #402
 - Use `function_selector!` to calculate transfer type selector in `Erc1155`. #417
-- Update internal functions of `Erc721` and `Erc721Consecutive` accept reference to `Bytes`. #437
+
+### Changed (Breaking)
+
+- Update internal functions of `Erc721` and `Erc721Consecutive` to accept a reference to `Bytes`. #437
+- Replace all `Call:new_in` to `Call:new` to stop supporting reentrancy. #440
```
Comment (0xNeshi) [inline CHANGELOG.md:30]:
```suggestion
- Stop supporting reentrancy, and borrow `self` immutably in `IErc721::_check_on_erc721_received` and `IErc1155::_check_on_erc1155_received`. #440
- Remove `storage: &mut impl TopLevelStorage` parameter from `ecdsa::recover`. #440
- Remove `TopLevelStorage` implementation from `VestingWallet`, `Erc1155`, `Erc20Permit`, `SafeErc20`, `Erc721Consecutive`, and `Erc721`. #440
```
Many things happened in this PR, let's list them all for awareness.

**Review by 0xNeshi — APPROVED** [review]
LGTM!

File: contracts/src/token/erc1155/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bidzyyys) [reply contracts/src/token/erc1155/mod.rs:783]: I like the idea.

[the remark this answers is not in the document]
Comment (bidzyyys) [reply contracts/src/token/erc1155/mod.rs:783]: If it was pointed out by `clippy` I am fine with this refactor.

File: CHANGELOG.md [hunk 7]
```diff
@@ -23,7 +23,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 - Implement `MethodError` for `safe_erc20::Error`. #402
 - Use `function_selector!` to calculate transfer type selector in `Erc1155`. #417
-- Update internal functions of `Erc721` and `Erc721Consecutive` accept reference to `Bytes`. #437
+
+### Changed (Breaking)
+
+- Update internal functions of `Erc721` and `Erc721Consecutive` to accept a reference to `Bytes`. #437
+- Stop supporting reentrancy, and borrow `self` immutably in `IErc721::_check_on_erc721_received` and `IErc1155::_check_on_erc1155_received`. #440
+- Remove `storage: &mut impl TopLevelStorage` parameter from `ecdsa::recover`. #440
+- Remove `TopLevelStorage` implementation from `VestingWallet`, `Erc1155`, `Erc20Permit`, `SafeErc20`, `Erc721Consecutive`, and `Erc721`. #440
```
Comment (bidzyyys) [inline CHANGELOG.md:32]:
```suggestion
- Remove `TopLevelStorage` trait implementation from `VestingWallet`, `Erc1155`, `Erc20Permit`, `SafeErc20`, `Erc721Consecutive`, and `Erc721`. #440
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -23,7 +23,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Implement `AddAssignUnchecked` and `SubAssignUnchecked` for `StorageUint`. #418
 - Implement `MethodError` for `safe_erc20::Error`. #402
 - Use `function_selector!` to calculate transfer type selector in `Erc1155`. #417
-- Update internal functions of `Erc721` and `Erc721Consecutive` accept reference to `Bytes`. #437
+
+### Changed (Breaking)
+
+- Update internal functions of `Erc721` and `Erc721Consecutive` to accept a reference to `Bytes`. #437
+- Stop supporting reentrancy, and borrow `self` immutably in `IErc721::_check_on_erc721_received` and `IErc1155::_check_on_erc1155_received`. #440
```
Comment (bidzyyys) [inline CHANGELOG.md:30]: Update following to the change suggested by `clippy`.

**Review by bidzyyys — COMMENTED** [review]
Apply clippy warning, update CHANGELOG properly and we are good to go 🚀

File: CHANGELOG.md (hunk 7 above)
Comment (SarveshLimaye) [reply CHANGELOG.md]: Done ✅

File: CHANGELOG.md (hunk 8 above)
Comment (SarveshLimaye) [reply CHANGELOG.md]: Applied changes suggested by `clippy` and updated `CHANGELOG` ✅

File: contracts/src/token/erc1155/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (SarveshLimaye) [reply contracts/src/token/erc1155/mod.rs]: Removed `self` from ` _check_on_erc1155_received` and made it an associated function. ✅

**Review by 0xNeshi — APPROVED** [review]
Very good job @SarveshLimaye !

Don't worry about the failing "Check Links" job, [it'll be handled](https://github.com/OpenZeppelin/rust-contracts-stylus/pull/450)

**Review by bidzyyys — APPROVED** [review]
Looks good!

**Review by 0xNeshi — APPROVED** [review]

**Review by qalisander — APPROVED** [review]

**Pull request merged**